internal/tasks/manifests.go

```diff
@@ -894,39 +900,84 @@ func (j *Janitor) CheckTrivySecurityStatusJob(registerer prometheus.Registerer)
 			},
 		},
 		BeginTx: j.db.Begin,
-		DiscoverRow: func(_ context.Context, tx *gorp.Transaction, _ prometheus.Labels) (securityInfo keppel.TrivySecurityInfo, err error) {
-			err = tx.SelectOne(&securityInfo, securityCheckSelectQuery, j.timeNow())
-			return securityInfo, err
+		DiscoverRow: func(_ context.Context, tx *gorp.Transaction, _ prometheus.Labels) (securityInfos []keppel.TrivySecurityInfo, err error) {
+			_, err = tx.Select(&securityInfos, securityCheckSelectQuery, j.timeNow())
+
+			// TODO: How to convince tx.Select to do this?
+			if len(securityInfos) == 0 {
+				err = sql.ErrNoRows
+			}
+
+			return securityInfos, err
 		},
 		ProcessRow: j.processTrivySecurityInfo,
 	}).Setup(registerer)
 }
 
-func (j *Janitor) processTrivySecurityInfo(ctx context.Context, tx *gorp.Transaction, securityInfo keppel.TrivySecurityInfo, labels prometheus.Labels) error {
-	//load corresponding repo, account and manifest
-	repo, err := keppel.FindRepositoryByID(tx, securityInfo.RepositoryID)
-	if err != nil {
-		return fmt.Errorf("cannot find repo for manifest %s: %w", securityInfo.Digest, err)
-	}
-	account, err := keppel.FindAccount(tx, repo.AccountName)
-	if err != nil {
-		return fmt.Errorf("cannot find account for repo %s: %w", repo.FullName(), err)
-	}
-	manifest, err := keppel.FindManifest(tx, *repo, securityInfo.Digest)
-	if err != nil {
-		return fmt.Errorf("cannot find manifest for repo %s and digest %s: %w", repo.FullName(), securityInfo.Digest, err)
+type returnStruct struct {
+	securityInfo keppel.TrivySecurityInfo
+	err          error
+}
+
+// processTrivySecurityInfo parallises the CheckTrivySecurityStatusJob jobloop without requiring extra database connections
+// It processes SecurityInfos in batches in the value of trivySecurityInfoBatchSize and runs the value of trivySecurityInfoThreads in parallel.
+func (j *Janitor) processTrivySecurityInfo(ctx context.Context, tx *gorp.Transaction, securityInfos []keppel.TrivySecurityInfo, labels prometheus.Labels) error {
+	// prevent deadlocks by waiting for more securityInfos in range below
+	lenSecurityInfo := len(securityInfos)
+	if trivySecurityInfoBatchSize < lenSecurityInfo {
+		trivySecurityInfoBatchSize = lenSecurityInfo
+	}
+
+	inputChan := make(chan keppel.TrivySecurityInfo, trivySecurityInfoBatchSize)
```

Comment:
Why is the batch size adjusted down here? Since it can only ever go down, it will tend to 0 over time. But since the SQL query still uses the original batch size, eventually the adjusted batch size will be too small and `inputChan` will not be able to hold all items.

Reply:
I converted this to a local variable.